sycl/include/CL/sycl/detail/spinlock.hpp

```diff
@@ -0,0 +1,39 @@
+//==-------------------- spinlock.hpp --- Spin lock ------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/sycl/detail/defines.hpp>
+
+#include <atomic>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+/// SpinLock is a synchronization primitive, that uses atomic variable and
+/// causes thread trying acquire lock wait in loop while repeatedly check if
+/// the lock is available.
+///
+/// One important feature of this implementation is that std::atomic<bool> can
+/// be zero-initialized. This allows SpinLock to have trivial constructor and
+/// destructor, which makes it possible to use it in global context (unlike
+/// std::mutex, that doesn't provide such guarantees).
+class SpinLock {
+public:
+  void lock() {
+    while (MLock.exchange(true, std::memory_order_acquire))
+      ;
```

Comment:
I think we might want want to "pause" here to avoid excessive power consumption in some CPU

Reply:
+1 from my side. 
BTW: is std::this_thread::yield(); good enouth r we need to use something else?

Reply:
That's what I mean: https://software.intel.com/content/www/us/en/develop/articles/a-common-construct-to-avoid-the-contention-of-threads-architecture-agnostic-spin-wait-loops.html

Reply:
[This](https://en.cppreference.com/w/cpp/atomic/atomic_flag) recommends using `atomic_flag` instead of `atomic_bool` as it is guaranteed to be lock-free.
+1 for yield

Reply:
[This](https://en.cppreference.com/w/cpp/atomic/atomic_flag) recommends using atomic_flag instead of atomic_bool as it is guaranteed to be lock-free.
+1 for yield

Reply:
Is `_mm_pause` recommented in https://software.intel.com/content/www/us/en/develop/articles/a-common-construct-to-avoid-the-contention-of-threads-architecture-agnostic-spin-wait-loops.html Intel-specific? Does not seem to be in C++ standard.

Reply:
>  is std::this_thread::yield(); good enouth or we need to use something else?

I think this is enough.

Reply:
I decided to use `std::this_thread::yield()`. Spin locks are not first-class citizens in our runtime. When their usage is more intense, we can use a more complicated approach from TBB. Also +1 on `atomic_flag`.